File: pagerank_hw/pgse.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <queue>
#include <tuple>
#include <algorithm>
#include <filesystem>
using namespace std;

namespace fs = std::filesystem;

// —— 全局配置 —— 
const double damping_factor       = 0.85;     // 跳转概率
const double epsilon    = 1e-6;     // 收敛阈值
const int    max_iter   = 100;      // 最大迭代次数
const int    block_size = 1000;     // 每个条带大小
const std::string data_file = "Data.txt";

// 读取边文件，统计 outdeg 并分条带 —— 
void create_edge_stripes(int N, int num_stripes) {
    // 打开所有条带输出流
    vector<ofstream> stripe_fs;
    stripe_fs.reserve(num_stripes);
    for(int i = 0; i < num_stripes; ++i) {
        stripe_fs.emplace_back("stripe_" + std::to_string(i) + ".txt");
    }
    // 读入边，写到对应条带
    ifstream fin(data_file);
    int u, v;
    while(fin >> u >> v) {
        int sid = v / block_size;
        stripe_fs[sid] << u << " " << v << "\n";
    }
    for(auto &ofs : stripe_fs) ofs.close();
}

// 初始化 PageRank 向量，按条带写磁盘 
void write_initial_r(int N, int num_stripes) {
    for(int sid = 0; sid < num_stripes; ++sid) {
        int start = sid * block_size;
        int end   = std::min((sid+1)*block_size, N);
        ofstream rf("r_stripe_" + std::to_string(sid) + ".txt");
        for(int i = start; i < end; ++i) {
            rf << (1.0 / N) << "\n";
        }
    }
}

//  外存条带 PageRank 迭代 
void external_stripe_pagerank(int N, int num_stripes,
                              const std::unordered_map<int,int> &outdeg) 
{
    for(int iter = 0; iter < max_iter; ++iter) {
        // 计算死节点泄漏量
        double leaked = 0.0;
        for(int sid = 0; sid < num_stripes; ++sid) {
            int start = sid * block_size;
            ifstream rf("r_stripe_" + std::to_string(sid) + ".txt");
            string line;
            int idx = 0;
            while(getline(rf, line)) {
                double val = std::stod(line);
                int node = start + idx;
                if (outdeg.find(node) == outdeg.end()) {
                    leaked += damping_factor * val;
                }
                ++idx;
            }
        }
        double leaked_share = leaked / N;
        double base = (1 - damping_factor) / N;

        double diff = 0.0;
        // 按目的条带 更新
        for(int vid = 0; vid < num_stripes; ++vid) {
            int vstart = vid * block_size;
            int vend   = min((vid+1)*block_size, N);
            int m      = vend - vstart;

            // 新条带缓冲
            vector<double> r_new(m, base + leaked_share);

            // 枚举所有源条带 uid
            for(int uid = 0; uid < num_stripes; ++uid) {
                int ustart = uid * block_size;
                int uend   = std::min((uid+1)*block_size, N);

                // 读旧 r_stripe_uid
                vector<double> r_block;
                {
                    ifstream uf("r_stripe_" + std::to_string(uid) + ".txt");
                    string l;
                    while(getline(uf,l)) {
                        r_block.push_back(std::stod(l));
                    }
                }
                // 读对应目的边 stripe_vid
                ifstream ef("stripe_" + to_string(vid) + ".txt");
                int uu, vv;
                while(ef >> uu >> vv) {
                    if (uu >= ustart && uu < uend) {
                        auto it = outdeg.find(uu);
                        if (it != outdeg.end()) {
                            int d = it->second;
                            int local_u = uu - ustart;
                            int local_v = vv - vstart;
                            r_new[local_v] += damping_factor * (r_block[local_u] / d);
                        }
                    }
                }
            }

            // 写入临时新条带并累加 diff
            string new_name = "r_new_stripe_" + std::to_string(vid) + ".txt";
            ofstream wf(new_name);
            for(double val : r_new) {
                wf << val << "\n";
            }
            // 计算 L1 差异
            {
                ifstream oldf("r_stripe_" + std::to_string(vid) + ".txt");
                ifstream newf(new_name);
                string lo, ln;
                while(getline(oldf,lo) && getline(newf,ln)) {
                    diff += fabs(stod(ln) - stod(lo));
                }
            }
        }

        // 用新文件替换旧条带
        for(int sid = 0; sid < num_stripes; ++sid) {
            fs::rename("r_new_stripe_" + std::to_string(sid) + ".txt",
                       "r_stripe_"     + std::to_string(sid) + ".txt");
        }

        cout << "Iter " << iter << " diff=" << diff << std::endl;
        if (diff < epsilon) {
            cout << "Converged.\n";
            break;
        }
    }
}

// —— k 路归并输出 Top-K —— 
void merge_top100(int N, int num_stripes, int K=100) {
    // 对每个条带内部排序并写入 sorted_sid.txt
    std::vector<std::string> sorted_files;
    for(int sid = 0; sid < num_stripes; ++sid) {
        int start = sid * block_size;
        vector<std::pair<double,int>> buf;
        ifstream rf("r_stripe_" + std::to_string(sid) + ".txt");
        string line;
        int idx = 0;
        while(getline(rf,line)) {
            buf.emplace_back(std::stod(line), start + idx);
            ++idx;
        }
        sort(buf.begin(), buf.end(),
                  [](auto &a, auto &b){ return a.first > b.first; });
        string tmp = "sorted_" + std::to_string(sid) + ".txt";
        ofstream of(tmp);
        for(auto &p : buf) {
            of << p.second << "\t" << p.first << "\n";
        }
        sorted_files.push_back(tmp);
    }

    // 4.2 k 路归并
    struct Item { double neg_score; int sid, node; };
    auto cmp = [](Item const &a, Item const &b){
        return a.neg_score > b.neg_score;
    };
    priority_queue<Item,vector<Item>,decltype(cmp)> heap(cmp);

    vector<std::ifstream> ifs;
    for(auto &fn : sorted_files) {
        ifs.emplace_back(fn);
    }
    // 初始化堆
    for(int sid = 0; sid < ifs.size(); ++sid) {
        string line;
        if (getline(ifs[sid], line)) {
            istringstream ss(line);
            int nid; double score;
            ss >> nid >> score;
            heap.push({-score, sid, nid});
        }
    }

    // 输出 Top-K
    ofstream out("res_stripe_ext.txt");
    for(int i = 0; i < K && !heap.empty(); ++i) {
        auto it = heap.top(); heap.pop();
        out << it.node << "\t" << -it.neg_score << "\n";
        // 读取该文件下一行
        string line;
        if (getline(ifs[it.sid], line)) {
            istringstream ss(line);
            int nid; double score;
            ss >> nid >> score;
            heap.push({-score, it.sid, nid});
        }
    }

    // 清理临时文件
    for(auto &fn: sorted_files) {
        fs::remove(fn);
    }
}

// —— 主流程 —— 
int main(){
    // 统计 outdeg & N
    unordered_map<int,int> outdeg;
    unordered_set<int> nodes;
    {
        ifstream fin(data_file);
        int u,v;
        while(fin >> u >> v){
            nodes.insert(u); nodes.insert(v);
            outdeg[u]++;
        }
    }
    int N = *max_element(nodes.begin(), nodes.end()) + 1;
    int num_stripes = std::ceil(double(N)/block_size);
    std::cout << "Total nodes = " << N
              << ", stripes = " << num_stripes << std::endl;

    create_edge_stripes(N, num_stripes);
    write_initial_r(N, num_stripes);
    external_stripe_pagerank(N, num_stripes, outdeg);
    merge_top100(N, num_stripes, 100);

    cout << "Done. Results in res_stripe_ext_cpp.txt\n";
    return 0;
}

File: pagerank_hw/pagerank大作业_details/000000_Python示例_000001_李华_第一次作业/源码/pgse.py
#!/usr/bin/env python3
import math
import os
import heapq

# 参数设置
beta = 0.85         # 随机游走时跟随超链接的概率，其余概率均匀跳转
epsilon = 1e-6      # 收敛阈值
max_iter = 100      # 最大迭代次数
block_size = 1000   # 条带宽度，根据内存和磁盘 I/O 调整

data_file = "Data.txt"

# 1. 读取边并分区到条带文件（按目标节点 v）
outdeg = {}         # 全局出度映射：node -> count
nodes = set()
with open(data_file) as f:
    for line in f:
        parts = line.strip().split()
        if len(parts) < 2:
            continue
        u, v = map(int, parts)
        nodes.add(u); nodes.add(v)
        outdeg[u] = outdeg.get(u, 0) + 1

N = max(nodes) + 1
num_stripes = math.ceil(N / block_size)

def create_edge_stripes():
    stripe_fs = [open(f"stripe_{i}.txt", "w") for i in range(num_stripes)]
    with open(data_file) as f:
        for line in f:
            parts = line.strip().split()
            if len(parts) < 2:
                continue
            u, v = map(int, parts)
            sid = v // block_size
            stripe_fs[sid].write(f"{u} {v}\n")
    for sf in stripe_fs:
        sf.close()

# 2. 初始化 PageRank：将 r 分块写入磁盘
def write_initial_r():
    for sid in range(num_stripes):
        start = sid * block_size
        end = min((sid + 1) * block_size, N)
        with open(f"r_stripe_{sid}.txt", "w") as rf:
            for _ in range(start, end):
                rf.write(f"{1.0/N}\n")

# 3. 外存化 PageRank 迭代

def external_stripe_pagerank():
    for iter_num in range(max_iter):
        # 计算流失
        leaked = 0.0
        for sid in range(num_stripes):
            start = sid * block_size
            with open(f"r_stripe_{sid}.txt") as rf:
                for local_idx, line in enumerate(rf):
                    i = start + local_idx
                    val = float(line)
                    if outdeg.get(i, 0) == 0:
                        leaked += beta * val
        leaked_share = leaked / N

        diff = 0.0
        for vid in range(num_stripes):
            v_start = vid * block_size
            v_end = min((vid + 1) * block_size, N)
            size_v = v_end - v_start
            r_new = [(1 - beta) / N + leaked_share for _ in range(size_v)]

            for uid in range(num_stripes):
                u_start = uid * block_size
                u_end = min((uid + 1) * block_size, N)
                with open(f"r_stripe_{uid}.txt") as uf:
                    r_block = [float(x) for x in uf]
                stripe_file = f"stripe_{vid}.txt"
                if not os.path.exists(stripe_file):
                    continue
                with open(stripe_file) as ef:
                    for line in ef:
                        u, v = map(int, line.split())
                        if u_start <= u < u_end:
                            out = outdeg.get(u, 0)
                            if out:
                                r_new[v - v_start] += beta * (r_block[u - u_start] / out)

            new_file = f"r_new_stripe_{vid}.txt"
            with open(new_file, "w") as wf:
                for val in r_new:
                    wf.write(f"{val}\n")
            with open(f"r_stripe_{vid}.txt") as oldf, open(new_file) as newf:
                for old_line, new_line in zip(oldf, newf):
                    diff += abs(float(new_line) - float(old_line))

        for sid in range(num_stripes):
            os.replace(f"r_new_stripe_{sid}.txt", f"r_stripe_{sid}.txt")
        if diff < epsilon:
            break

# 4. 合并各分块并流式输出 Top100
def merge_top100(output_file="res_stripe_ext.txt", k=100):
    # 先对每个条带排序并写入临时文件
    sorted_files = []
    for sid in range(num_stripes):
        start = sid * block_size
        scores = []
        with open(f"r_stripe_{sid}.txt") as rf:
            for idx, line in enumerate(rf):
                scores.append((float(line), start + idx))
        scores.sort(reverse=True)
        tmp = f"sorted_{sid}.txt"
        with open(tmp, "w") as sf:
            for score, nid in scores:
                sf.write(f"{nid}\t{score}\n")
        sorted_files.append(tmp)

    # k 路归并写 Top k
    files = [open(f, "r") for f in sorted_files]
    heap = []
    for sid, f in enumerate(files):
        line = f.readline().strip()
        if not line: continue
        nid, score = line.split("\t")
        heapq.heappush(heap, (-float(score), sid, int(nid)))

    with open(output_file, "w") as out:
        for _ in range(k):
            if not heap: break
            neg, sid, nid = heapq.heappop(heap)
            out.write(f"{nid}\t{-neg}\n")
            line = files[sid].readline().strip()
            if line:
                nid2, score2 = line.split("\t")
                heapq.heappush(heap, (-float(score2), sid, int(nid2)))

    for f in files:
        f.close()
    for tmp in sorted_files:
        os.remove(tmp)

# 主流程
if __name__ == '__main__':
    create_edge_stripes()
    write_initial_r()
    external_stripe_pagerank()
    merge_top100()
    print("External-stripe PageRank top100 wrote to res_stripe_ext.txt")
